README.txt
Texturepacker
=============

Minecraft Texture-Pack Maker

Purpose
-------

Texturepacker provides program ``maketexture`` for assembling
texture packs for Minecraft using recipes written in a simple language
expressed using JSON or as Python dicts, and a library of Python classes
that can be used to do the same thing in your own programs.

Texture packs can be created by combining image files in PNG format, or
images extracted from existing texture packs. This enables you to remix
texture packs to use alternative textures for some elements if the
original artist included them, or even to combine textures from two or
more separate packs. One use for this might be to upgrade a pack
designed for an earlier version of Minecraft to work with newly added
items.

Installation
------------

Requirements:

- Python 2.7
- Distribute (or Setuptools might work)

You should be able to install with the following command::

    $ python setup.py install

You can test it worked by running it on an example recipe::

    $ maketexture -v examples/groovystipple.zip

(These are included in the source distribution but will not be
included in a binary install.)

To run automated tests, you need to install 2 more packages and
then use Nose::

    $ pip install nose mock
    $ nosetests tests

This should print a lot of dots and then report OK at the end.


NEWS.txt
This file outlines changes to the package, in reverse chronological order.

NOTE. The version number appears in 3 files:
- This file
- setup.py
- maketexture

0.12 (2012-02-04)
====

- Update Minecraft atlass for Release 1.2
- Add Planet Minecraft to repertoire of the unwrapper

0.11.1
====

- Bug fix for texture-pack URLs containing a query-string part.

0.11
====

- Packs downloaded with HTTP URLs will be unwrapped first
  (this allows for indirect URLs using some hosting services)

0.10
====

- Added map updates for Minecraft 1.0
- Recipes can create textures starting with a blank image instead of an existing source.

0.9.1
=====

- Added map updates for Minecraft Beta 1.7.
- Split some of the maps files out of the examples directory to create `minecraft` and `altpacks`

0.8.4
=====

19 June 2011

- Allow override of the default loader when creating a `Mixer` instance.

0.8.3
=====

24 April 2011

- Bug fix: Don’t throw exception when source pack has single-line `pack.txt`.

0.8.2
=====

24 April 2011

- Bug fix: when unjumbling a texture pack, don’t change its label and
  description to ‘unjumbled’.

0.8.1
=====

12 April 2011

- Bug fix: composite images using palletized source images had ugly
  results because the replaced times would be force-mapped on to the
  base image’s colour pallet. Images are now uplifted to RGBA format to
  prevent this.

0.8
===

8 April 2011

- `MapBase.get_css` expresses a tile from a texture as CSS styles.

0.7
===

4 April 2011

- `Loader.add_local_knowledge` allows you to use a local copy of HTTP
  resources (should be useful on the web site that happens to implement
  those resources)

0.6.3 (Head cold special)
========================

3 April 2011


- `Map.get_alt_list` allows `stone1` as well as `stone_1`, but knows
  that `crop0`, ..., `crop7` are not alternative versions of a
  nonexistent texture `crop`.

- Bug fix: using a string to fetch an atlas resulted in an infinite
  loop! Now the string is instead treated as a URL

0.6
===

2 April 2011

- Feature: map method `Map.get_alt_list` gets a list of candiates for
  making an alternative-textures remix of a pack.

0.5.1
=====

10 March 2011

- Bug fix: Setting a cache directory with `set_http_cache` should now work!

0.5
====

10 March 2011

- New recipe keywords `unjumble` and `if_missing` to work with ZIPs containing files with
  the correct names but in the wrong folders.

0.4.1
=====

9 March 2011

- Bug fix: LazyPack returned wrong value for last-modified. Missing test :-(

0.4
===

9 March 2011

- New recipe keyword: `pack_icon`: create a `pack.png` from the terrain file.

- Add `get_last_modified` and `is_modified_since` methods to resources and
  packs.

- Hence make the `maketexture` script avoid rebuilding packs if their
  ingredients are unmodified.

- Support `../` and `./` when resolving partial references with custom URLs
  as base URL per <http://tools.ietf.org/html/rfc3986>.

0.3
===

27 February 2011

- Loader has `add_scheme` method so you can hook in to internal recipe
  database by defining custom URL schemes.

0.2
===

27 February 2011

- Recipes can declare formal pack parameters which must be filled in to
  make the recipe.

- Labels and description in recipes can depend on labels and
  descriptions of constituent packs.

- Added support for Minecraft Beta 1.3

0.1.1
=====

23 February 2011

Initial release to PyPI

tests/test_unwrapper.py
# -*-coding: UTF-8-*-

"""Given the alleged URL of a texture pack, find plausing forum and download URLs.

This is tricky because often they are embedded in a page that is wrapped in
a bit.ly link which is wrapped in an adf.ly link and the download is hosted on Mediafire
behind yeat another layer of redirects."""

import unittest
from mock import Mock, patch

import sys
import os
import httplib2
import json
from texturepacker.unwrapper import *

# First some helper functions for faking the HTTP requests the tests involve …
#
# Generally the approach is to use httplib2 in Python to
# request the URL in question; the response is saved as JSON
# and the body as HTML in a file with the same name apart from
# the .json or .html extension.
#
# Then when writing the tests, an HttpStub instance is created
# and the resources linked to URLs by (file)name.

def get_data(file_name):
    """Return the raw data."""
    file_path = os.path.join(os.path.dirname(__file__), 'test_data/unwrapper', file_name)
    with open(file_path, 'rb') as strm:
        return strm.read()

def get_json(file_name):
    """Return the JSON-encoded data."""
    file_path = os.path.join(os.path.dirname(__file__),  'test_data/unwrapper', file_name)
    with open(file_path, 'rb') as strm:
        return json.load(strm)


class UnexpectedHttpRequest(Exception): pass

class HttpStub(object):
    """Pretends to be an httplib2.Http instance."""
    def __init__(self):
        self.responses = {}
        self.requests = []
        self.follow_redirects = True

    def add(self, url, name):
        self.responses[url] = name

    def request(self, url, *args, **kwargs):
        """Called by the code being tested.

        If the URL has been added with `add` then returns the
        appropriate response. Otherwise gets KeyError.
        """
        try:
            name = self.responses[url]
        except KeyError:
            raise UnexpectedHttpRequest('{0}: unexpected HTTP request'.format(url))
        self.requests.append((name, kwargs.get('headers'), self.follow_redirects))
        resp = get_json('%s.json' % name)
        body = get_data('%s.html' % name)
        return resp, body

def stub_http(*args):
    """Decorator to set up a temporary fake HTTP client library.

    Arguments --
        Tuples (URL, NAME) where NAME is used to find data files to
        use as the result of requesting the URL.

    Returns --
        A function that decorates the function this decorator is decorating.
    """
    http = HttpStub()
    for url, name in args:
        http.add(url, name)
    http_class = Mock()
    http_class.return_value = http
    inner_decorator = patch('httplib2.Http', http_class)
    def outer_decorator(func):
        def decorated_func(*args, **kwargs):
            args += (http,)
            return inner_decorator(func)(*args, **kwargs)
        return decorated_func
    return outer_decorator


class TestStubbity(unittest.TestCase):
    # Check that the stubbing itself is effective.
    @stub_http(('http://example.org/foo/bar/baz/quux', 'adfly'))
    def test_adfly_stubbed(self, http):
        http = httplib2.Http()
        resp, body = http.request('http://example.org/foo/bar/baz/quux')
        self.assertTrue('http://adf.ly/favicon.ico' in body)
        self.assertEqual('http://adf.ly/380075/forestdepths', resp['content-location'])
        self.assertEqual(['adfly'], http.requests)


# Tests for the individual sites’ unwrappers.
# These tests will need to be updated as they change their site code.

class TestHelperUnwrappers(unittest.TestCase):
    def test_unwrap_adfly(self):
        urls = unwrap_adfly('http://adf.ly/380075/forestdepths', get_json('adfly.json'), get_data('adfly.html'))
        self.assertEqual('http://bit.ly/pXTHAp', urls['next'])

    def test_unwrap_bitly(self):
        resp, body = get_json('forum1.json'), get_data('forum1.html')
        urls = unwrap_bitly('http://bit.ly/pXTHAp', resp, body)
        self.assertEqual(('http://www.minecraftforum.net/topic/617455-16x-18-forest-depths-wip-v04/', resp, body),
            urls['next'])

    def xtest_unwrap_mediafire(self):
        resp, body = get_json('mediafire.json'), get_data('mediafire.html')
        urls = unwrap_mediafire('http://www.mediafire.com/?p6gbi987u93t6os', resp, body)

        expected = 'http://www.mediafire.com/dynamic/download.php?qk=p6gbi987u93t6os&pk1=f8cff2a113114097978837db48750c2f0dbe1ae019c327dd15760ee74e3cb0aa93ed8fc77d41bccfedcc9f367b3597dc&r=3p0y3'
        self.assertEqual(expected, urls['next'])

    def xtest_unwrap_mediafire_download(self):

        resp, body = get_data('mediafire-download-2.json'), get_data('mediafire-download-2.html')
        urls = unwrap_mediafire_download('http://www.mediafire.com/dynamic/download.php?qk=xxx&pk1=xxx&r=xxx', resp, body)

        # This is my best guess based on poking through the JavaScript code!
        expected = 'http://download197.mediafire.com/010eb3678beg/p6gbi987u93t6os/ForestDepths+v0.4.zip'
        self.assertEqual(expected, urls['next'])

    def xtest_unwrap_mediafire_2011_11_29(self):
        # Abortive attempt to update now mediafie have re-obfuscated their code.
        resp, body = get_json('mediafire-11bt-1.json'), get_data('mediafire-11bt-1.html')
        urls = unwrap_mediafire('http://www.mediafire.com/?p6gbi987u93t6os', resp, body)

        self.assertTrue('next' in urls)
        # XXX expected resulkt?


class TestPlanetMinecraftUnwrapper(unittest.TestCase):
    def setUp(self):
        resp, body = get_json('planetminecraft-1.json'), get_data('planetminecraft-1.html')
        u = 'http://www.planetminecraft.com/texture_pack/the-clean-lines-texture-pack-357/'
        self.urls = unwrap_planetminecraft(u, resp, body)

    def test_home_url(self):
        self.assertEqual('http://www.planetminecraft.com/texture_pack/the-clean-lines-texture-pack-357/' ,
                self.urls['home'])

    def test_forum_url(self):
        self.assertEqual('http://www.minecraftforum.net/topic/96610-16x18clean-lines-texture-pack-v182/',
                self.urls['forum'])

    def test_download(self):
        self.assertEqual('http://www.planetminecraft.com/texture_pack/the-clean-lines-texture-pack-357/',
                self.urls['download'])

    def test_next(self):
        self.assertEqual('http://www.planetminecraft.com/texture_pack/the-clean-lines-texture-pack-357/download/file/470877/',
                self.urls['next'])


class TestPlantMinecraftDownloadUnwrapper(unittest.TestCase):
    def setUp(self):
        resp, body = get_json('planetminecraft-2.json'), get_data('planetminecraft-2.html')
        u = 'http://www.planetminecraft.com/texture_pack/the-clean-lines-texture-pack-357/download/file/470877/'
        self.urls = unwrap_planetminecraft_download(u, resp, body)

    def test_next(self):
        # Main gotcha here is that the redirect is to an invalid URL
        # (it contains spaces) which must be rewritten to work.
        self.assertEqual('http://cdn.planetminecraft.com/files/resource_media/texture/Clean%20Lines%20v182.zip',
                self.urls['next'])


class TestMinecraftforumUnwrapper(unittest.TestCase):
    def setUp(self):
        resp, body = get_json('forum1.json'), get_data('forum1.html')
        self.urls = unwrap_minecraftforum('http://www.minecraftforum.net/topic/617455-16x-18-forest-depths-wip-v04/', resp, body)

    def test_forum_url(self):
        self.assertEqual('http://www.minecraftforum.net/topic/617455-16x-18-forest-depths-wip-v04/', self.urls['forum'])

    def test_next_url(self):
        self.assertEqual('http://www.mediafire.com/?p6gbi987u93t6os', self.urls['next'])

class TestMinecraftforumz(unittest.TestCase):
    def do_unwrap(self, name, url=None):
        if not url:
            url = 'http://example.com/%s' % name
        resp, body = {'content-location': url}, get_data('%s.html' % name)
        urls = unwrap_minecraftforum(url, resp, body)
        return urls

    def test_bitly_address_with_image_label(self):
        urls = self.do_unwrap('forum2', 'http://www.minecraftforum.net/topic/584177-16x18-%e2%80%a2%e2%80%a2%e2%80%a2%e2%96%ba-comfy-n-cozy-%e2%97%84%e2%80%a2%e2%80%a2%e2%80%a2/')
        self.assertEqual('http://bit.ly/nUClBn', urls['next'])

    def test_adfly_address_with_text_label(self):
        urls = self.do_unwrap('forum3', 'http://www.minecraftforum.net/topic/572115-16x16-olive-173/')
        self.assertEqual('http://adf.ly/2fjAh', urls['next'])

    def test_alicence(self):
        urls = self.do_unwrap('forum3', 'http://www.minecraftforum.net/topic/572115-16x16-olive-173/')
        self.assertEqual('http://creativecommons.org/licenses/by-nc-nd/3.0/', urls['licence'])

    def test_adfly_address_with_text_label(self):
        # In this case it is really the home page not the download page.
        urls = self.do_unwrap('forum4')
        self.assertEqual('http://fl.xrbpowered.com/', urls['next'])


class TestGuessUrls(unittest.TestCase):
    def test_negative(self):
        self.assertEqual(0, guess_url_is_download('http://creativecommons.org/licenses/by-nc-nd/3.0/'))

    def test_mediafire(self):
        self.assertTrue(guess_url_is_download('http://www.mediafire.com/?p6gbi987u93t6os'))

    def test_dot_zip(self):
        self.assertTrue(guess_url_is_download('http://www.example.com/foo/nbar/baz.zip'))

    def test_mediafire(self):
        self.assertTrue(guess_url_is_home('http://www.planetminecraft.com/texture_pack/leostereos-textures-revamped/'))


# Test for taking a URL through several hops to its final source.
class TestUnwrapper(unittest.TestCase):
    @stub_http(('http://adf.ly/380075/forestdepths', 'adfly'), ('http://bit.ly/pXTHAp', 'forum1'), ('http://www.mediafire.com/?p6gbi987u93t6os', 'mediafire'), ('http://www.mediafire.com/dynamic/download.php?qk=p6gbi987u93t6os&pk1=f8cff2a113114097978837db48750c2f0dbe1ae019c327dd15760ee74e3cb0aa93ed8fc77d41bccfedcc9f367b3597dc&r=3p0y3', 'mediafire-download-2'))
    def setUp(self, http):
        self.http = http
        self.http.requests = []
        self.unwrapper = Unwrapper()
        self.urls = self.unwrapper.unwrap('http://adf.ly/380075/forestdepths')

    def test_checked_them_all(self):
        self.assertEqual(['adfly', 'forum1', 'mediafire', 'mediafire-download-2'], [x for (x, y, z) in self.http.requests])

    def test_found_forum(self):
        self.assertEqual('http://www.minecraftforum.net/topic/617455-16x-18-forest-depths-wip-v04/', self.urls['forum'])

    def test_found_download(self):
        expected = 'http://www.mediafire.com/?p6gbi987u93t6os'
        self.assertEqual(expected, self.urls['download'])

    def test_found_final_url(self):
        expected = 'http://download197.mediafire.com/010eb3678beg/p6gbi987u93t6os/ForestDepths+v0.4.zip'
        self.assertEqual(expected, self.urls['final'])

    def test_copied_cookie(self):
        # I don’t think the Mediafire download will work without it!
        self.assertTrue('ukey=c6keq7fw96q395gyxlyfg19ht1bj7o8s' in self.http.requests[-1][1]['cookie'])


# Test for upwrapping a URL only enough to find the download URL
class TestPartialUnwrapper(unittest.TestCase):
    @stub_http(('http://adf.ly/380075/forestdepths', 'adfly'), ('http://bit.ly/pXTHAp', 'forum1'))
    def setUp(self, http):
        self.http = http
        self.http.requests = []
        self.unwrapper = Unwrapper()
        self.urls = self.unwrapper.unwrap('http://adf.ly/380075/forestdepths', until=['download', 'forum'])

    def test_checked_them_all(self):
        self.assertEqual(['adfly', 'forum1'], [x for (x, y, z) in self.http.requests])

    def test_found_forum(self):
        self.assertEqual('http://www.minecraftforum.net/topic/617455-16x-18-forest-depths-wip-v04/', self.urls['forum'])

    def test_found_download(self):
        expected = 'http://www.mediafire.com/?p6gbi987u93t6os'
        self.assertEqual(expected, self.urls['download'])

    def test_not_found_final_url(self):
        self.assertTrue('final' not in self.urls)


# Test for upwrapping a URL only enough to find the download URL
class TestPlanetMinecraftDownload(unittest.TestCase):
    @stub_http(
        ('http://www.planetminecraft.com/texture_pack/the-clean-lines-texture-pack-357/', 'planetminecraft-1'),
        ('http://www.planetminecraft.com/texture_pack/the-clean-lines-texture-pack-357/download/file/470877/', 'planetminecraft-2'))
    def setUp(self, http):
        self.http = http
        self.http.requests = []
        self.unwrapper = Unwrapper()
        self.urls = self.unwrapper.unwrap('http://www.planetminecraft.com/texture_pack/the-clean-lines-texture-pack-357/')

    def test_final_is_download(self):
        self.assertEqual('http://cdn.planetminecraft.com/files/resource_media/texture/Clean%20Lines%20v182.zip', self.urls['final'])

    def test_redirections_were_off(self):
        name, headers, follow_redirects = self.http.requests[-1]
        self.assertEqual(False, follow_redirects)


if __name__ == '__main__':
    sys.exit(unittest.main())









src/texturepacker/unwrapper.py
# -*-coding: UTF-8-*-

"""Given the alleged URL of a texture pack, find plausible forum and download URLs.

This is tricky because often they are embedded in a page that is wrapped in
a bit.ly link which is wrapped in an adf.ly link and the download is hosted on Mediafire
behind yet another layer of redirects.
"""

import sys
import os
import httplib2
import re
from BeautifulSoup import BeautifulSoup
import json
from urlparse import urljoin

# Unwrapper functions take a URL, response, and body
# (these being the result of a request for that URL)
#  and optional Http object
# and return a dict of categorized URLs or None.

# The following decoratots allow us to specify which
# URL patterns the functions relate to.

def unwrapper_for(url_pattern, follow_redirects=True):
    """Decorator for unwrapper functions.

    Arguments --
        url_pattern -- regex for matching URLs that this applies to.
            NOTE. The initial http:// and www. will have already been stripped.
        follow_redirects -- set this to false to not follow redirects
    """
    if isinstance(url_pattern, basestring):
        url_pattern = re.compile(url_pattern)
    def func_wrapper(func):
        func.url_pattern = url_pattern
        func.follow_redirects = follow_redirects
        return func
    return func_wrapper


ADFLY_RE = re.compile(ur"""
    function \s close_bar\(\) \s \{ \s+
    self\.location \s = \s '(?P<next>[^']*)'; \s+
    \}
    """, re.VERBOSE)
@unwrapper_for(r'^adf\.ly')
def unwrap_adfly(url, resp, body, http=None):
    # adf.ly wraps the target page in a frameset or something
    # It has a close button which restores the target page.
    m = ADFLY_RE.search(body)
    if m:
        return m.groupdict()

@unwrapper_for(r'^bit\.ly')
def unwrap_bitly(url, resp, body):
    # Redirects happen transparently
    # We just need to tell caller to process the results using the next unwrapper.
    return {
        'next': (resp['content-location'], resp, body),
    }

@unwrapper_for(r'^planetminecraft\.com/texture_pack/[a-z-]+-\d+/$')
def unwrap_planetminecraft(url, resp, body):
    """Planet Minecraft texture-pack section.

    Attempt to fish out the genuine download and forum URLs linked therefrom"""
    urls = {
        'home': resp['content-location'],
        'download': resp['content-location'],
    }
    soup = BeautifulSoup(body)
    table_elt = soup.first('div', 'resource-share').table
    for tr_elt in table_elt.findAll('tr'):
        a_elt = tr_elt.td.a
        label = a_elt['title']
        href = a_elt['href']
        if label == 'Download Texture Pack':
            urls['next'] = urljoin(url, href)
        elif label == 'Visit Forum post':
            urls['forum' ] = href
    return urls

@unwrapper_for(r'^planetminecraft\.com/texture_pack/[a-z-]+-\d+/download/file/\d+/$',
        follow_redirects=False)
def unwrap_planetminecraft_download(url, resp, body):
    """The dowload link from Planet Minecraft"""
    # The maon gotcha is that they redirecy to an invalid URL
    # (it contains spaces)!
    return {
        'next': resp['location'].replace(' ', '%20')
    }


URL_SCORES = [
    (re.compile(r'^http://(www\.)?mediafire\.com/'), 50),
    (re.compile(r'^http://bit\.ly/'), 25),
    (re.compile(r'^http://adf\.ly/'), 25),
]
LABEL_SCORES = [
    (re.compile('download', re.IGNORECASE), 100),
]
LICENCE_SCORES = [
    (re.compile('^http://creativecommons.org/licenses/'), 100),
]
WOT_NO_SLASH_RE = re.compile(r'^https?://[\w.-]+$')

@unwrapper_for(r'^minecraftforum\.net')
def unwrap_minecraftforum(url, resp, body):
    urls = {
        'forum': url, # This *is* the forum page!
    }

    # It might also have clues as to where the downloads are.
    soup = BeautifulSoup(body)
    post_elt = soup.first('div', 'entry-content')

    best_href = None
    best_score = 0
    for a_elt in post_elt.findAll('a', 'bbc_url'): # This class distinguishes URLS inserted by the author of the post.
        try:
            href = a_elt['href']

            # Many entries contain self-links.
            if href == url:
                continue

            # Check for licence link.
            licence_score = sum(pat_score for (pat, pat_score) in LICENCE_SCORES if pat.search(href))
            if licence_score:
                urls['licence'] = href
                continue

            # Otherwise, this is a candidate for download or home link.
            score = sum(pat_score for (pat, pat_score) in URL_SCORES if pat.search(href))
            labels = []
            try:
                label = ''.join(a_elt.findAll(text=True))
                if label:
                    labels.append(label)
            except AttributeError:
                pass
            # Look for label immediately preceeding link:
            label = a_elt.findPreviousSibling(text=True)
            if label:
                labels.append(label)

            for label in labels:
                score += sum(pat_score for (pat, pat_score) in LABEL_SCORES if pat.search(label))
            if a_elt.img:
                score += 10
            if score > best_score:
                best_href, best_score = href, score
        except KeyError, e:
            print >>sys.stderr, a_elt, 'did not have', e
    if best_href:
        if WOT_NO_SLASH_RE.match(best_href):
            best_href += '/'
        urls['next'] = best_href

    return urls

MEDIAFIRE_PKR_RE = re.compile(r"^<!--\s*var LA=\s*false;\s*pKr='([^']+)';")
MEDIAFIRE_CALL_RE = re.compile(r'^DoShow\("notloggedin_wrapper"\);\s*cR\(\);\s*(\w+)\(\);')
MEDIAFIRE_CALL2_RE = re.compile(r"\w+\('\w+','([a-f\d]+)'\)")

@unwrapper_for(r'^mediafire\.com/\?')
def unwrap_mediafire(url, resp, body):
    """Mediafire wrap downloads in a mass of obsfucated JavaScript. Extract and return the next URL."""
    urls = {
        'download': url,
    }

    soup = BeautifulSoup(body)

    # The first of three codes needed to generate the next URL.
    secret_qk = url.split('?', 1)[1]

    # The JavaScript calls a function whose name is randomized.
    # So our first job is to find the name of the function.
    for elt in soup.body.findAll('script', type='text/javascript', language=None):
        s = elt.string
        s = s and s.strip()
        if s:
            m = MEDIAFIRE_CALL_RE.search(s)
            if m:
                secret_func = m.group(1)
                break

    # That function has within it some 'encrypted' code
    # which in turn contains the third key.
    FUNC_PAT = re.compile(r"""
        function \s %s .*
        unescape\('(?P<cyphertext>[a-f\d]+)'\); \s*
        \w+ = (?P<count>\d+) ; \s*
        for \( .* \) .*
        \w+ = \w+ \+ \(String.fromCharCode\(parseInt\(\w+.substr\(i\s\*\s2,\s2\),\s16\) \^
                (?P<key>[\d^]+)
        \)\); \s*
        eval\(\w+\);
        """ % secret_func, re.VERBOSE)

    for elt in soup.body.findAll('script', type='text/JavaScript', language='JavaScript'): # lol capitalization
        m = MEDIAFIRE_PKR_RE.search(elt.string)
        if m:
            # This is the second code needed to get the next URL.
            secret_pKr = m.group(1)

            # Now to find the 'encrypted' function and pull the third code out of it.
            m = FUNC_PAT.search(elt.string)
            if m:
                plaintext = mediafire_decode(m.group('cyphertext'), m.group('count'), m.group('key'))

                # Now hoik the third code out of the plaintext.
                m = MEDIAFIRE_CALL2_RE.match(plaintext)
                secret_pk1 = m.group(1)
            break

    # The next URL is also on Mediafire -- this is the dynamically generated page
    # that is loaded in to an invisible IFRAME.
    urls['next'] = ('http://www.mediafire.com/dynamic/download.php?qk=%s&pk1=%s&r=%s'
                % (secret_qk, secret_pk1, secret_pKr))
    return urls

def mediafire_decode(cyphertext, count, key):
    """Mediafire use a simple cypher to obscure some codes embedded in their HTML.

    Arguments --
        cyphertext -- lowercase hexadecimal digits. Mediafire needlessly wrap it in unescape(...).
        count -- number of pairs of digits to consider; a int or string representation of an int.
        key -- the expression XORed with all bytes, in the form of a string
                containing ^-separated integers. For example, "13^7".

    Returns --
        A byte string, usually containg JavaScript code to pass to eval.
    """
    # Super-secred decryption in progress!!
    count = int(count)
    key = reduce(lambda x, y: x ^ y, [int(x) for x in key.split('^')])
    plaintext = ''.join(chr(int(cyphertext[2 * i:2 * i + 2], 16) ^ key) for i in range(count))
    return plaintext

MF_DOWNLOAD_ENIGMA_RE = re.compile(r"""
    unescape\('(?P<cyphertext>[a-f\d]+)'\);
    \w+ = (?P<count>\d+) ;
    for \(i = 0; i < \w+; i\+\+ \)
    \w+ = \w+ \+ \(String.fromCharCode\(parseInt\(\w+.substr\(i\s\*\s2,\s2\),\s16\) \^
            (?P<key>[\d^]+)
    \)\);
    eval\(\w+\); """, re.VERBOSE)
MF_DOWNLOAD_ROTOR_RE = re.compile(r"(\w+)='(\w+)';")
MF_DOWNLOAD_INNERHTML_RE = re.compile(r"""
    case \s 15:  # Ensure we have the correct branch of the switch
    .*?
    href=\\"(?P<prefix>http://download\d+.mediafire.com/)"
    \s* \+ \s*
    (?P<key>\w+)
    \s* \+ \s*
    "(?P<suffix>[^"]+\.zip)\\"
    """, re.VERBOSE)


@unwrapper_for(r'^mediafire\.com/dynamic/download\.php\?')
def unwrap_mediafire_download(url, resp, body):
    """Pull out the link to the actual download from Medafire’s hidden page"""

    # This only works if the cookies and other stuff are set up correctly.
    # There is only one SCRIPT element so no point in anlysing the HTML structure …
    m = MF_DOWNLOAD_ENIGMA_RE.search(body)
    if m:
        code = mediafire_decode(m.group('cyphertext'), m.group('count'), m.group('key'))
        enigma = dict(MF_DOWNLOAD_ROTOR_RE.findall(code))
        m = MF_DOWNLOAD_INNERHTML_RE.search(body)
        href = m.group('prefix') + enigma[m.group('key')] + m.group('suffix')
        return {
            'next': href,
        }
    return {}

@unwrapper_for(r'.*')
def unwrap_anything_and_save_it(url, resp, body):
    """For debugging! This unwrapper saves its arguments to disk for later analysis."""
    name = url.split('/')[2].replace('www.', '')
    with open('%s.json' % name, 'wb') as strm:
        json.dump(resp, strm, indent=4)
    print >>sys.stderr, 'Wrote headers to %s.json' % name
    with open('%s.html' % name, 'wb') as strm:
        strm.write(body)
    print >>sys.stderr, 'Wrote body to %s.html' % name
    return {}


URL_IS_DOWNLOAD_SCORES = [
    (re.compile('^http://(www\.)?mediafire.com/', re.IGNORECASE), 50),
    (re.compile('\.zip$', re.IGNORECASE), 50),
]
def guess_url_is_download(url):
    """Return a positive number if this URL looks likely to be a ZIP download."""
    return sum(score for (pat, score) in URL_IS_DOWNLOAD_SCORES if pat.search(url))

URL_IS_HOME_SCORES = [
    (re.compile('^http://(www\.)?planetminecraft\.com/texture_pack/[\w-]+/', re.IGNORECASE), 100),
]
def guess_url_is_home(url):
    """Return a positive number if this URL looks likely to be an external homepage for the pack.."""
    return sum(score for (pat, score) in URL_IS_HOME_SCORES if pat.search(url))



# DRIVER

UNWRAPPERS = [
    unwrap_planetminecraft,
    unwrap_planetminecraft_download,
    unwrap_adfly,
    unwrap_bitly,
    unwrap_minecraftforum,
    unwrap_mediafire,
    unwrap_mediafire_download,
]
COOKIE_EXPIRES_RE = re.compile("""
    expires=
    (Mon|Tue|Wed|Thu|Fri|Sat|Sun), \s [^;,]+
    [;,]?
    """, re.VERBOSE)

class Unwrapper(object):
    """Device for peeling back layers of redirection web sites to get the actual download link.

    The `unwrap` methodis used to attempt to find download etc. URLs.

    """
    def __init__(self, http=None):
        self.http = http or httplib2.Http()

    def unwrap(self, url, until=None):
        """Try to find actual download and forum URLs, starting with this one.

        Arguments --
            url (string) -- the published URL for the resource (may be the actual URL
                or the address of a forum page linking to it etc.)
            until (Set of string) -- stop when all of the named categorized URLs are known.
                Use this when you want to document the resource
                but not necessarly find the final one.

        Returns --
            A dictionary  with some or none of the following keys defined:

            final -- the URL that could not be further processed; generally
                if all has gone well this will be the actual URL of the resource itself.
            forum -- a forum post concerning this resource (this happens when
                the alleged URL of the resource is actually a link to the post).
            download -- the URL to quote as the one to download the resource from
                (may still involve a redirection via Mediafire etc.).
            home -- the URL to quote as the home page of the resource

        If `until` is not specified, then carry on until there is no more
        dereferencing to do.

        This may make one or more HTTP requests, and will consume
        memory as it holds the intermediate resources for examination.
        """
        if until:
            until = set(until)

        cookie_jar = {}
        queue = [url]
        result = {}
        while queue:
            url_resp_body = queue.pop(0)
            need_download = isinstance(url_resp_body, basestring)
            if need_download:
                url = url_resp_body
            else:
                url, resp, body = url_resp_body
            x = url
            if x.startswith('http://'):
                x = x[7:]
            if x.startswith('www.'):
                x = x[4:]
            for func in UNWRAPPERS:
                if func.url_pattern.search(x):
                    # Have found a handler for this URL.
                    if need_download:
                        # Dereference the URL; also takes care of copying cookies beetween requests.
                        headers = {}
                        self.http.follow_redirects = func.follow_redirects if hasattr(func, 'follow_redirects') else True
                        if cookie_jar:
                            headers['cookie'] = ';'.join('%s=%s' % (key, val) for (key, val) in cookie_jar.items())
                        resp, body = self.http.request(url, headers=headers)
                        if 'set-cookie' in resp:
                            # Highly simplified cookie-wrangling—we assume we just want to copy all of them.
                            cookies_line = COOKIE_EXPIRES_RE.sub('', resp['set-cookie'])
                            for cookie_def in cookies_line.split(','):
                                parts = cookie_def.strip().split(';')
                                cookie_name, cookie_value = parts[0].split('=', 1)
                                cookie_jar[cookie_name] = cookie_value
                    res = func(url, resp, body)

                    if res and 'forum' in res and 'next' in res:
                        # We have found a forum page; its links might be download or home.
                        # Only the obvious cases are handled here (e.g., *.zip is a download).
                        # More subtle cases will require a further download.
                        u = res['next']
                        if 'download' not in res and guess_url_is_download(u):
                            res['download'] = u
                        if 'home' not in res and guess_url_is_home(u):
                            res['home'] = u

                    result.update(res)
                    break
            else:
                # Could not unwrap this URL, so we assume it is the final source.
                result['final'] = url
            if until and not (until - set(result.keys())):
                break
            if 'next' in result:
                queue.append(result.pop('next'))
        return result

default_unwrapper = None

def unwrap(*args, **kwargs):
    """"Shortcut for unwrapping a URL.

    Equivalent to Unwrapper().unwrap(...).
    """
    global default_unwrapper
    if not default_unwrapper:
        default_unwrapper = Unwrapper()
    return default_unwrapper.unwrap(*args, **kwargs)

if __name__ == '__main__':
    import pprint
    u = 'http://adf.ly/380075/forestdepths'
    u = 'http://www.planetminecraft.com/texture_pack/the-clean-lines-texture-pack-357/'
    pprint.pprint(Unwrapper().unwrap(u))

